src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -335,199 +297,172 @@ Build all the hash tables for the partitions of the build column. One job per pa
-      if (chunk_id < chunk_offsets.size() - 1) {
-        input_size = chunk_offsets[chunk_id + 1] - input_offset;
-      }
-
-      for (size_t chunk_offset = input_offset; chunk_offset < input_offset + input_size; ++chunk_offset) {
-        const auto& element = container_elements[chunk_offset];
+  for (ChunkID input_partition_idx{0}; input_partition_idx < input_partition_count; ++input_partition_idx) {
+    jobs.emplace_back(std::make_shared<JobTask>([&, input_partition_idx]() {
+      const auto& input_partition = radix_container[input_partition_idx];
+      for (auto input_idx = size_t{0}; input_idx < input_partition.elements.size(); ++input_idx) {
+        const auto& element = input_partition.elements[input_idx];
 
         // In case of NULL-removing inner-joins, we ignore all NULL values.
         // Such values can be created in several ways: join input already has non-phyiscal NULL values (non-physical
         // means no RowID, e.g., created during an OUTER join), a physical value is NULL but is ignored for an inner
         // join (hence, we overwrite the RowID with NULL_ROW_ID), or it is simply a remainder of the pre-sized
         // RadixPartition which is initialized with default values (i.e., NULL_ROW_IDs).
-        if (!retain_null_values && element.row_id == NULL_ROW_ID) {
-          continue;
+        if (!retain_null_values) {
+          Assert(!(element.row_id == NULL_ROW_ID), "NULL_ROW_ID should not have made it this far");
         }
 
-        const size_t radix = hash_function(static_cast<HashedType>(element.value)) & mask;
+        const size_t radix = hash_function(static_cast<HashedType>(element.value)) & radix_mask;
+
+        auto& output_idx = output_offsets_by_input_partition[input_partition_idx][radix];
+        Assert(output_idx < output[radix].elements.size(), "output_idx is completely out-of-bounds");
+        if (input_partition_idx < input_partition_count - 1) {
+          Assert(output_idx < output_offsets_by_input_partition[input_partition_idx + 1][radix],
+                 "output_idx goes into next range");
+        }
 
-        // In case NULL values have been materialized in materialize_input(),
-        // we need to keep them during the radix clustering phase.
+        // In case NULL values have been materialized in materialize_input(), we need to keep them during the radix
+        // clustering phase.
         if constexpr (retain_null_values) {
-          (*output_nulls)[output_offsets[radix]] = null_value_bitvector[chunk_offset];
+          null_values_as_char[radix][output_idx] = input_partition.null_values[input_idx];
         }
 
-        (*output)[output_offsets[radix]] = element;
-        ++output_offsets[radix];
+        output[radix].elements[output_idx] = element;
+
+        ++output_idx;
       }
     }));
     jobs.back()->schedule();
   }
+  Hyrise::get().scheduler()->wait_for_tasks(jobs);
+  jobs.clear();
 
+  // Compress null_values_as_char into partition.null_values
```

Comment:
Is this also necessary when `retain_null_values` is false?

Reply:
No, added a check